### PR TITLE
Allows token-syncer to update previously soft-deleted tokens

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- do not require etag while updating soft-deleted token
- requires etags for hard delete of active tokens

## Why are we making these changes?

We would like the token syncer to be able to sync tokens.

### CODE REVIEW

**Comment by sradack** [comment]
Can we explain why this was a problem?  On the surface, it seems it might be correct that we would not stick soft-deleted tokens in the previous metadata.

**Comment by shamsimam** [comment]
You're right, it seems correct that a soft-deleted token should not show up in the previous metadata. I've changed the syncer to handle updating soft-deleted tokens.

File: waiter/src/waiter/token.clj [hunk 1]
```diff
@@ -236,15 +236,11 @@
           (let [token-owner (get token-metadata "owner")
                 version-hash (get headers "if-match")]
             (if hard-delete
-              (do
-                (when-not version-hash
```
Comment (sradack) [inline waiter/src/waiter/token.clj]: Are we still ensuring that hard-deletes aren't happening on non-soft-deleted tokens?

Comment (shamsimam) [reply waiter/src/waiter/token.clj]: Did we used to ensure it in the past? Afaik, any admin can hard-delete any token. 
The token syncer does hard deletes only if the token is soft-deleted on all clusters.

Comment (sradack) [reply waiter/src/waiter/token.clj]: What allows us to remove this constraint?  Why don't we want `If-Match` on hard-deletes?

Comment (sradack) [reply waiter/src/waiter/token.clj]: If we are hard deleting, require `If-Match` if not soft-deleted.

**Comment by shamsimam** [comment]
Ready for review.

File: waiter/integration/waiter/token_request_test.clj [hunk 2]
```diff
@@ -428,14 +428,6 @@
                   (assert-response-status response 412)
                   (is (str/includes? (str body) "Cannot modify stale token"))))
 
-              (testing "hard-delete without etag"
```
Comment (sradack) [inline waiter/integration/waiter/token_request_test.clj]: Can we keep this test, or is there a change in behavior here?

Comment (shamsimam) [reply waiter/integration/waiter/token_request_test.clj]: Added back the integration test.

**Pull request merged**